firebase-firestore/src/main/java/com/google/firebase/firestore/local/BundleSerializer.java

```diff
@@ -0,0 +1,383 @@
+    int totalDocuments = bundleMetadata.getInt("totalDocuments");
+    long totalBytes = bundleMetadata.getLong("totalBytes");
+    return new BundleMetadata(bundleId, version, createTime, totalDocuments, totalBytes);
+  }
+
+  public BundledDocumentMetadata decodeBundledDocumentMetadata(String json) throws JSONException {
+    JSONObject object = new JSONObject(json);
+    return decodeBundledDocumentMetadata(object);
+  }
+
+  public BundledDocumentMetadata decodeBundledDocumentMetadata(JSONObject bundledDocumentMetadata)
+      throws JSONException {
+    DocumentKey key = DocumentKey.fromPath(decodeName(bundledDocumentMetadata.getString("name")));
+    SnapshotVersion readTime =
+        decodeSnapshotVersion(bundledDocumentMetadata.getJSONObject("readTime"));
+    boolean exists = bundledDocumentMetadata.optBoolean("exists", false);
+    JSONArray queriesJson = bundledDocumentMetadata.optJSONArray("queries");
+    List<String> queries = new ArrayList<>();
+    // Technically, the queries array should never be missing, but we treat a missing and empty
+    // array the same to avoid crashing during data import.
+    if (queriesJson != null) {
+      for (int i = 0; i < queriesJson.length(); ++i) {
+        queries.add(queriesJson.getString(i));
+      }
+    }
+    return new BundledDocumentMetadata(key, readTime, exists, queries);
+  }
+
+  @VisibleForTesting
+  Document decodeDocument(String json) throws JSONException {
+    JSONObject object = new JSONObject(json);
+    return decodeDocument(object);
+  }
+
+  Document decodeDocument(JSONObject document) throws JSONException {
+    String name = document.getString("name");
+    DocumentKey key = DocumentKey.fromPath(decodeName(name));
+    SnapshotVersion updateTime = decodeSnapshotVersion(document.getJSONObject("updateTime"));
+
+    Value.Builder value = Value.newBuilder();
+    decodeMapValue(value, document.getJSONObject("fields"));
+
+    return new Document(
+        key,
+        updateTime,
+        ObjectValue.fromMap(value.getMapValue().getFieldsMap()),
+        Document.DocumentState.SYNCED);
+  }
+
+  private ResourcePath decodeName(String name) {
+    ResourcePath resourcePath = ResourcePath.fromString(name);
+    if (!remoteSerializer.isLocalResourceName(resourcePath)) {
+      throw new IllegalArgumentException(
+          "Resource name is not valid for current instance: " + name);
+    }
+    return resourcePath.popFirst(5);
+  }
+
+  private SnapshotVersion decodeSnapshotVersion(JSONObject timestamp) {
+    long seconds = timestamp.optLong("seconds", 0);
+    int nanos = timestamp.optInt("nanos", 0);
+    return new SnapshotVersion(new Timestamp(seconds, nanos));
+  }
+
+  private Query decodeQuery(JSONObject bundledQuery) throws JSONException {
+    JSONObject structuredQuery = bundledQuery.getJSONObject("structuredQuery");
+
+    ResourcePath parent = decodeName(bundledQuery.getString("parent"));
+    verifyNoSelect(structuredQuery);
```

Comment:
This seems like it could be paired with the read of the `structuredQuery` (similar to the way `from` is validated immediately after reading it).